src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodFilteringService.cs

```diff
@@ -0,0 +1,191 @@
+                {
+                    // Alway create indices for documents in current project if they don't exist.
+                    loadOnly &= projectId != currentProject.Id;
+
+                    var project = solution.GetProject(projectId);
+                    if (project == null || !project.SupportsCompilation)
+                    {
+                        continue;
+                    }
+
+                    // Transitively get all the PE references
+                    peReferences.AddRange(project.MetadataReferences.OfType<PortableExecutableReference>());
+                    foreach (var document in project.Documents)
+                    {
+                        // Don't look for extension methods in generated code.
+                        if (document.State.Attributes.IsGenerated)
+                        {
+                            continue;
+                        }
+
+                        var info = await document.GetSyntaxTreeIndexAsync(loadOnly, cancellationToken).ConfigureAwait(false);
+
+                        // Don't provide anyting if we don't have all the required SyntaxTreeIndex created.
+                        if (info == null)
+                        {
+                            return null;
+                        }
+
+                        if (info.ContainsExtensionMethod)
+                        {
+                            syntaxTreeIndexBuilder.Add(info);
+                        }
+                    }
+                }
+
+                foreach (var peReference in peReferences)
+                {
+                    var info = await SymbolTreeInfo.GetInfoForMetadataReferenceAsync(
+                        solution, peReference, loadOnly, cancellationToken).ConfigureAwait(false);
+
+                    // Don't provide anyting if we don't have all the required SymbolTreeInfo created.
+                    if (info == null)
+                    {
+                        return null;
+                    }
+
+                    if (info.ContainsExtensionMethod)
+                    {
+                        symbolTreeInfoBuilder.Add(info);
+                    }
+                }
+
+                var results = new MultiDictionary<string, string>();
+
+                // Find matching extension methods from source.
+                foreach (var info in syntaxTreeIndexBuilder)
+                {
+                    // Add simple extension methods with matching target type name
+                    foreach (var targetTypeName in targetTypeNames)
+                    {
+                        if (!info.SimpleExtensionMethodInfo.TryGetValue(targetTypeName, out var methodInfoIndices))
+                        {
+                            continue;
+                        }
+
+                        foreach (var index in methodInfoIndices)
+                        {
+                            if (info.TryGetDeclaredSymbolInfo(index, out var methodInfo))
```

Comment:
can this actuall fail if given an index in methodInfoIndices?  if not, have an assert/contract.

Reply:
shouldn't be possible. Fixed